state-reconstruct-fetcher/src/types/v3.rs

```diff
@@ -90,60 +95,126 @@ impl TryFrom<&abi::Token> for V3 {
             number_of_l1_txs,
             priority_operations_hash,
             system_logs,
-            total_l2_to_l1_pubdata,
+            pubdata_commitments,
         };
 
         Ok(blk)
     }
 }
 
+impl V3 {
+    pub async fn parse_pubdata(
+        &self,
+        client: &BlobHttpClient,
+    ) -> Result<Vec<L2ToL1Pubdata>, ParseError> {
+        let bytes = &self.pubdata_commitments[..];
+        match self.pubdata_source {
+            PubdataSource::Calldata => {
+                let l = bytes.len();
+                if l < CALLDATA_SOURCE_TAIL_SIZE {
+                    Err(ParseError::InvalidCalldata("too short".to_string()))
+                } else {
+                    parse_resolved_pubdata(&bytes[..l - CALLDATA_SOURCE_TAIL_SIZE])
+                }
+            }
+            PubdataSource::Blob => parse_pubdata_from_blobs(bytes, client).await,
+        }
+    }
+}
+
 // Read the source of the pubdata from a byte array.
 fn parse_pubdata_source(bytes: &[u8], pointer: &mut usize) -> Result<PubdataSource, ParseError> {
     let pubdata_source = u8::from_be_bytes(read_next_n_bytes(bytes, pointer));
     pubdata_source.try_into()
 }
 
-fn parse_total_l2_to_l1_pubdata(
+async fn parse_pubdata_from_blobs(
     bytes: &[u8],
-    pointer: &mut usize,
-    pubdata_source: PubdataSource,
+    client: &BlobHttpClient,
 ) -> Result<Vec<L2ToL1Pubdata>, ParseError> {
-    match pubdata_source {
-        PubdataSource::Calldata => parse_pubdata_from_calldata(bytes, pointer),
-        PubdataSource::Blob => todo!(),
+    let mut pointer = 0;
+    let mut l = bytes.len();
+    let mut blobs = Vec::new();
+    while pointer < l {
+        let pubdata_commitment = &bytes[pointer..pointer + PUBDATA_COMMITMENT_SIZE];
+        let blob = get_blob(&pubdata_commitment[48..96], client).await?;
+        let mut blob_bytes = ethereum_4844_data_into_zksync_pubdata(&blob);
+        blobs.append(&mut blob_bytes);
+        pointer += PUBDATA_COMMITMENT_SIZE;
     }
-}
 
-fn parse_pubdata_from_calldata(
-    bytes: &[u8],
-    pointer: &mut usize,
-) -> Result<Vec<L2ToL1Pubdata>, ParseError> {
-    let mut l2_to_l1_pubdata = Vec::new();
+    l = blobs.len();
+    while l > 0 && blobs[l - 1] == 0u8 {
+        l -= 1;
+    }
 
-    // Skip over logs and messages.
-    let num_of_l1_to_l2_logs = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
-    *pointer += L2_TO_L1_LOG_SERIALIZE_SIZE * num_of_l1_to_l2_logs as usize;
+    let blobs_view = &blobs[..l];
+    parse_resolved_pubdata(blobs_view)
+}
 
-    let num_of_messages = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
-    for _ in 0..num_of_messages {
-        let current_message_len = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
-        *pointer += current_message_len as usize;
+async fn get_blob(kzg_commitment: &[u8], client: &BlobHttpClient) -> Result<Vec<u8>, ParseError> {
```

Comment:
We should not pass `BlobHttpClient` to this function. The function itself should rather be a part of `BlobHttPClient`.

Reply:
I don't think it makes a particular difference, but OK...